tools/block_cache_trace_analyzer.cc

```diff
@@ -0,0 +1,325 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "tools/block_cache_trace_analyzer.h"
+
+#include <cinttypes>
+#include <set>
+#include "monitoring/histogram.h"
+
+namespace rocksdb {
+namespace {
+std::string block_type_to_string(TraceType type) {
+  switch (type) {
+    case kBlockTraceFilterBlock:
+      return "Filter";
+    case kBlockTraceDataBlock:
+      return "Data";
+    case kBlockTraceIndexBlock:
+      return "Index";
+    case kBlockTraceRangeDeletionBlock:
+      return "RangeDeletion";
+    case kBlockTraceUncompressionDictBlock:
+      return "UncompressionDict";
+    default:
+      break;
+  }
+  // This cannot happen.
+  assert(false);
+  return "";
+}
+
+std::string caller_to_string(BlockCacheLookupCaller caller) {
+  switch (caller) {
+    case kUserGet:
+      return "Get";
+    case kUserMGet:
+      return "MultiGet";
+    case kUserIterator:
+      return "Iterator";
+    case kPrefetch:
+      return "Prefetch";
+    case kCompaction:
+      return "Compaction";
+    default:
+      break;
+  }
+  // This cannot happen.
+  assert(false);
+  return "";
+}
+}  // namespace
+
+BlockCacheTraceAnalyzer::BlockCacheTraceAnalyzer(
+    const std::string& trace_file_path)
+    : trace_file_path_(trace_file_path) {
+  env_ = rocksdb::Env::Default();
+}
+
+void BlockCacheTraceAnalyzer::RecordAccess(
+    const BlockCacheTraceRecord& access) {
+  ColumnFamilyStats& cf_stats = cf_stats_map_[access.cf_name];
+  SSTFileStats& file_stats = cf_stats.fd_stats_map[access.sst_fd_number];
+  file_stats.level = access.level;
+  BlockTypeStats& block_type_stats =
+      file_stats.block_type_stats_map[access.block_type];
+  BlockStats& block_stats = block_type_stats.block_stats_map[access.block_key];
+  block_stats.AddAccess(access);
+}
+
+Status BlockCacheTraceAnalyzer::Analyze() {
+  std::unique_ptr<TraceReader> trace_reader;
+  Status s =
+      NewFileTraceReader(env_, EnvOptions(), trace_file_path_, &trace_reader);
+  if (!s.ok()) {
+    return s;
+  }
+  BlockCacheTraceReader reader(std::move(trace_reader));
+  s = reader.ReadHeader(&header_);
+  if (!s.ok()) {
+    return s;
+  }
+  while (s.ok()) {
+    BlockCacheTraceRecord access;
+    s = reader.ReadAccess(&access);
+    if (!s.ok()) {
+      break;
```

Comment:
If s is not ok here and we break the loop, we should return this status instead of ok in the end of this function.

Reply:
Sure. I updated to return 's'.